rules/prefer-error-cause.js

```diff
@@ -0,0 +1,388 @@
+/* eslint-disable capitalized-comments */
+/* eslint-disable brace-style */
+'use strict';
+const esquery = require('esquery');
+const {isCommaToken} = require('eslint-utils');
+const _ = require('lodash');
+const {matches, methodCallSelector} = require('./selectors/index.js');
+const {appendArgument, replaceReferenceIdentifier} = require('./fix/index.js');
+const getReferences = require('./utils/get-references.js');
+
+const ERROR = 'error';
+const SUGGESTION = 'suggestion';
+
+const messages = {
+	[ERROR]: 'Prefer specifying an old error as `cause` option where rethrowing the error.',
+	[SUGGESTION]: 'Specify `cause` option to Error\'s constructor.',
+};
+
+const catchClauseSelector = matches([
+	[
+		'CatchClause',
+		':has(ThrowStatement)',
+	].join(''),
+]);
+
+const promiseCatchSelector = matches([
+	[
+		methodCallSelector({method: 'catch', argumentsLength: 1}),
+		':has(ThrowStatement)',
+	].join(''),
+]);
+
+const peekTop = array => {
+	if (_.isEmpty(array)) {
+		return undefined;
+	}
+
+	return array[array.length - 1];
+};
+
+const insertProperty = (fixer, node, text, sourceCode) => {
+	const [penultimateToken, lastToken] = sourceCode.getLastTokens(node, 2);
+	if (node.properties.length > 0) {
+		text = isCommaToken(penultimateToken) ? `${text}` : `, ${text}`;
+	}
+
+	return fixer.insertTextBefore(lastToken, text);
+};
+
+const generateCauseInspector = inspectProperty => objectExpression => {
+	if (!objectExpression || objectExpression.type !== 'ObjectExpression') {
+		return false;
+	}
+
+	return objectExpression.properties.some(property =>
+		property.key.name === 'cause' && inspectProperty(property),
```

Comment:
```suggestion
		!property.computed && property.key.name === 'cause' && inspectProperty(property),
```

```js
const cause = 'not-cause'
new Error('', {[cause]: 'foo'});
```

We should check `...` element here.

```js
new Error('', {...foo});
```

We can't know if `cause` is in `foo`.